src/category_theory/limits/shapes/constructions/finite_products_of_binary_products.lean

```diff
@@ -0,0 +1,211 @@
+  end,
+  fac' := λ s,
+  begin
+    rintro ⟨j⟩,
+    apply fin.induction_on j,
+    { apply (binary_fan.is_limit.lift' t₂ _ _).2.1 },
+    { rintro i -,
+      dsimp only [extend_fan_π_app],
+      rw [fin.cases_succ, ← assoc, (binary_fan.is_limit.lift' t₂ _ _).2.2, t₁.fac],
+      refl }
+  end,
+  uniq' := λ s m w,
+  begin
+    apply binary_fan.is_limit.hom_ext t₂,
+    { rw (binary_fan.is_limit.lift' t₂ _ _).2.1,
+      apply w ⟨0⟩ },
+    { rw (binary_fan.is_limit.lift' t₂ _ _).2.2,
+      apply t₁.uniq ⟨_, _⟩,
+      rintro ⟨j⟩,
+      rw assoc,
+      dsimp only [discrete.nat_trans_app],
+      rw ← w ⟨j.succ⟩,
+      dsimp only [extend_fan_π_app],
+      rw fin.cases_succ }
+  end }
+
+section
+variables [has_binary_products.{v} C] [has_terminal C]
+
+/--
+If `C` has a terminal object and binary products, then it has a product for objects indexed by
+`ulift (fin n)`.
+-/
+lemma has_product_ulift_fin :
+  Π (n : ℕ) (f : ulift (fin n) → C), has_product f
+| 0 := λ f,
+  begin
+    letI : has_limits_of_shape (discrete (ulift (fin 0))) C :=
+      has_limits_of_shape_of_equivalence
+        (discrete.equivalence (equiv.ulift.trans fin_zero_equiv').symm),
+    apply_instance,
+  end
+| (n+1) := λ f,
+  begin
+    haveI := has_product_ulift_fin n,
+    apply has_limit.mk ⟨_, extend_fan_is_limit f (limit.is_limit _) (limit.is_limit _)⟩,
+  end
+
+/--
+If `C` has a terminal object and binary products, then it has limits of shape
+`discrete (ulift (fin n))` for any `n : ℕ`.
+-/
+lemma has_limits_of_shape_ulift_fin (n : ℕ) :
+  has_limits_of_shape (discrete (ulift (fin n))) C :=
+{ has_limit := λ K,
+begin
+  letI := has_product_ulift_fin n K.obj,
+  let : discrete.functor K.obj ≅ K := discrete.nat_iso (λ i, iso.refl _),
+  apply has_limit_of_iso this,
+end }
+
+/-- If `C` has a terminal object and binary products, then it has finite products. -/
+lemma has_finite_products_of_has_binary_and_terminal : has_finite_products C :=
+λ J 𝒥₁ 𝒥₂,
+begin
+  resetI,
+  rcases fintype.equiv_fin J with ⟨e⟩,
+  apply has_limits_of_shape_of_equivalence (discrete.equivalence (e.trans equiv.ulift.symm)).symm,
+  refine has_limits_of_shape_ulift_fin (fintype.card J),
+end
+end
+
+section preserves
+variables (F : C ⥤ D)
+variables [preserves_limits_of_shape (discrete walking_pair) F]
+variables [preserves_limits_of_shape (discrete pempty) F]
+variables [has_finite_products.{v} C]
+
+/--
+If `F` preserves the terminal object and binary products, then it preserves products indexed by
+`ulift (fin n)` for any `n`.
+-/
+noncomputable def preserves_fin_of_preserves_binary_and_terminal  :
+  Π (n : ℕ) (f : ulift (fin n) → C), preserves_limit (discrete.functor f) F
+| 0 := λ f,
+  begin
+    letI : preserves_limits_of_shape (discrete (ulift (fin 0))) F :=
+      preserves_limits_of_shape_of_equiv
+        (discrete.equivalence (equiv.ulift.trans fin_zero_equiv').symm) _,
+    apply_instance,
+  end
+| (n+1) :=
+  begin
```

Comment:
I'm still scared by definitions that use such a huge `begin...end` block. But I guess universal properties hide all the pain?

Reply:
Yeah that's reasonable, but `preserves_limit` is subsingleton again, so there shouldn't be any pain